### PR TITLE
[DataGrid] Fix render

### LINKED ISSUE #1243: [DataGrid] Last row cut with autoHeight=true, density="comfortable"
If you set autoHeight = true and density = "comfortable" the last row will not be fully visible (using 27 rows).

`@material-ui/data-grid@4.0.0-alpha.22`

### LINKED ISSUE #1249: [DataGrid] Issue with rendering
occurred between version alpha-20 => 21
https://material-ui.com/components/data-grid/rendering/#customization-example

The scrollbar is not working.

### PR DESCRIPTION
**Pull request by dtassone**
Test with 
https://deploy-preview-1253--material-ui-x.netlify.app/storybook/?path=/story/x-grid-tests-dataset--grid-x-rows-play

fix #1249, 
fix #1243 

- [x] add tests

### CODE REVIEW

File: packages/grid/_modules_/grid/models/gridContainerProps.ts [hunk 1]
```diff
@@ -24,6 +24,10 @@ export type GridViewportSizeState = ElementSize;
  * The set of container properties calculated on resize of the grid.
  */
 export interface GridContainerProps {
+  /**
+   * if true, the grid is virtualizing the rendering of rows.
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridContainerProps.ts:28]:
```suggestion
   * If `true`, the grid is virtualizing the rendering of rows.
```

File: packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx [hunk 2]
```diff
@@ -171,6 +171,14 @@ describe('<DataGrid /> - Keyboard', () => {
     expect(getActiveCell()).to.equal('4-0');
   });
 
+  it('Space only should go to the bottom of the page even with small number of rows', () => {
+    render(<KeyboardTest nbRows={4} />);
+    getCell(0, 0).focus();
+    expect(getActiveCell()).to.equal('0-0');
+    fireEvent.keyDown(document.activeElement!, SPACE_KEY);
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:178]: I personally think it would be much better with:

```suggestion
    fireEvent.keyDown(document.activeElement!, { key: ' ' });
```

It optimizes for duplication of information and explicitlness. You will find 18 matches in the core with `, { key: ' ' }`.

Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:178]: Also, it's weird that we have `code` (unused) or `shiftKey` (default) defined in:

https://github.com/mui-org/material-ui-x/blob/76948930e3a9fa6d0d03fe60c54b60d3a60e2202/packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx#L20

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx [hunk 3]
```diff
@@ -10,7 +10,9 @@ import {
   GridRowData,
   useGridApiRef,
   XGrid,
+  XGridProps,
 } from '@material-ui/x-grid';
+import { useData } from '../../../../storybook/src/hooks/useData';
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:15]: Makes moving and reading code easier:

```suggestion
import { useData } from 'packages/storybook/src/hooks/useData';
```

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx [hunk 4]
```diff
@@ -249,4 +251,97 @@ describe('<XGrid /> - Rows ', () => {
...
+      expect(isVirtualized).to.equal(true);
+    });
+
+    it('should render last row when scrolling to the bottom', () => {
+      const { container } = render(<TestCaseVirtualization nbRows={996} hideFooter height={600} />);
+      const totalHeight = apiRef!.current!.getState().containerSizes?.totalSizes.height;
+
+      const gridWindow = container.querySelector('.MuiDataGrid-window');
+      const renderingZone = container.querySelector('.MuiDataGrid-renderingZone');
+      gridWindow.scrollTop = totalHeight;
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:322]: How about we use the same approach as @DanailH https://github.com/mui-org/material-ui-x/blob/76948930e3a9fa6d0d03fe60c54b60d3a60e2202/packages/grid/x-grid/src/tests/events.XGrid.test.tsx#L269

But to hedge any potential issues with (12345 isn't enough) 

```suggestion
      gridWindow.scrollTop = 10e6;
```

This way, we can refactor the internal of the grid without having to care about `apiRef!.current!.getState().containerSizes?.totalSizes.height`. We scroll to the bottom, end of the story.

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx [hunk 5]
```diff
@@ -249,4 +251,97 @@ describe('<XGrid /> - Rows ', () => {
...
+      isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(true);
+    });
+
+    it('should render last row when scrolling to the bottom', () => {
+      const { container } = render(<TestCaseVirtualization nbRows={996} hideFooter height={600} />);
+      const totalHeight = apiRef!.current!.getState().containerSizes?.totalSizes.height;
+
+      const gridWindow = container.querySelector('.MuiDataGrid-window');
+      const renderingZone = container.querySelector('.MuiDataGrid-renderingZone');
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:321]: I recall @eps1lon encouraging to go all-in on the: "one component is rendered at a time in the page, no need to scope to the container" approach.

```suggestion
      const gridWindow = document.querySelector('.MuiDataGrid-window');
      const renderingZone = document.querySelector('.MuiDataGrid-renderingZone');
```

x the other cases

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx [hunk 6]
```diff
@@ -249,4 +251,97 @@ describe('<XGrid /> - Rows ', () => {
...
+      gridWindow.scrollTop = totalHeight;
+      gridWindow.dispatchEvent(new Event('scroll'));
+
+      const lastCell = container.querySelector(
+        '.MuiDataGrid-row:last-child .MuiDataGrid-cell:first-child',
```
Comment (oliviertassinari) [inline packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:326]: Could we always use the a11y features first (the role instead of the class names)?
```suggestion
        '[role="row"]:last-child [role="cell"]:first-child',
```

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts [hunk 7]
```diff
@@ -121,14 +130,21 @@ export const useGridContainerProps = (
...
+      if (options.autoPageSize || options.autoHeight || !isVirtualized) {
+        // we don't need vertical virtualization in these cases.
+        const viewportPageSize =
+          options.autoHeight || !isVirtualized
+            ? rowsCount
+            : Math.floor(viewportSizes.height / rowHeight);
+        const requiredHeight = viewportPageSize * rowHeight;
 
         const indexes: GridContainerProps = {
+          isVirtualized,
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: It looks like a leaky abstraction. Isn't virtualization supposed to only be a concern of `useGridVirtualRows.ts`? It seems that it's only computed here is it can be tested. This sounds like the opposite of an end-to-end test. Won't these tests be slow to update when we want to refactor the implementation?
```suggestion
```

**Review by oliviertassinari — COMMENTED** [review]
The two bugs seems fixed.

Thanks for the extra test cases. Considering that the bug seems related to the size of the scrollbar (at least it's a variable to taken into account), writing tests that can are enough end-to-end to be resilient to future refactoring about accurate enough to really test something sounds like a true challenge.

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts (hunk 7 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: I wonder if always keeping the virtualization on could be simpler.

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx (hunk 4 above)
Comment (dtassone) [reply packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:322]: Yes but if you put a random value you're not testing anything anymore. What I test here is that the calculated value corresponds to the end of the page

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts (hunk 7 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: You're not disabling when it suits. It is on by default when more than 2 rows are outside the viewport. So we can keep a natural experience when we have a small set of rows.

File: packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx (hunk 2 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:178]: When you fire a keydown event you have code and shiftKey, we have different behavior for space+shift.
Also if you start using code in your implementation, your test won't work anymore.

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx (hunk 4 above)
Comment (oliviertassinari) [reply packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:322]: > test here is that the calculated value corresponds to the end of the page

Then we can rewrite the expect to explicitly test it. It's not obvious. You could for instance assert that the scroll height is equal or set a really high scroll top and assert that the browsers max it out to the correct value?

File: packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx (hunk 2 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/keyboard.DataGrid.test.tsx:178]: This was my point, code shouldn't be used, shift is already the default.

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts (hunk 7 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: I don't follow what you mean my natural experience. If there is no scrollbar, there is no virtulization, by definition. It still seems thar we have the DOM structure in place for virtulization when is Virtualize is false.

In any case, my main point of interpretation was about leaking a value to the public API that doesn't seem to solve any problem for developers, and to have it in place that doesn't seem related to virtualization (hence my question about the leaky abstraction).

Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: This value is in the grid state like all the other values that the grid needs. It is used in other places.
Why is it not related to virtualisation?

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx (hunk 4 above)
Comment (dtassone) [reply packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:322]: I don't understand why it's confusing, not sure about what you mean, how it will be a better test?

I assign totalHeight to the scroll top value so window goes down to the totalHeight value. 
Then I try to retrieve the last cell, so if the virtualisation or the totalHeight is wrong then I won't find it.
Finally I assert that the rendering zone has the expected transform value.

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts (hunk 7 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: > Why is it not related to virtualisation?

I meant that `isVirtualized` is computed in `useGridContainerProps.ts` while it's only used in `useGridVirtualRows.ts`. It seems that we could compute and consuming it in `useGridVirtualRows.ts`, as a locally scoped variable. This would also have the advantage of making it impossible to test in the end-to-end tests (so we can focus on the behavior instead).

File: packages/grid/x-grid/src/tests/rows.XGrid.test.tsx (hunk 4 above)
Comment (oliviertassinari) [reply packages/grid/x-grid/src/tests/rows.XGrid.test.tsx:322]: I propose that we break`gridWindow.scrollTop = totalHeight;` down. As you said, this line serves two purposes:

1. It's meant to scroll to the bottom of the scroll container, to make sure that even the last cell is rendered
2. It's meant to make sure state.totalHeight is actually the total height.

For 1. I would propose:

```jsx
gridWindow.scrollTop = 10e6; // scroll to the bottom
```

no ambiguity possible about the objective 

For 2. I would propose:

```jsx
expect(gridWindow.scrollHeight).to.equal(totalHeight);
```

so we can also detect if `totalHeight` is too tall (no only too small). When the code assigns the `element.scrollTop` property, the browser do `Math.min(element.scrollHeight, new value)`.

File: packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts (hunk 7 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: it's used in 2 places

https://github.com/mui-org/material-ui-x/pull/1253/files#diff-7904e8c986f79f8bbc1627613039ba9fc46cc3f000aef6b4883c56f4aff2311aR154

https://github.com/mui-org/material-ui-x/pull/1253/files#diff-b067e2d006d18d8c1214e487d80d2a267f8bc6e8b163055fb0c381317db7010fR138

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts:147]: My bad, I didn't notice it it was read too.

I didn't look at the root issue, so I can only share my high-level perspective on the problem. It seems to go in the opposite direction to where we should aim:

1. It seems to melt `useGridVirtualRows` and `useGridContainerProps` together. Shouldn't we aim for more isolation?
2. It seems to create more branching rendering behavior. The more we have, the more challenging pinning, dynamic rows, tree data will be to handle. Shouldn't we aim for fixing the virtualization edge cases at their core?

In any case, the bugs are fixed, and we have a couple of new constraints that should help for the next time we iterate on the problem 👍  (the tests that depends on isVirtualized are unlikely to be very helpful in the future, but we can easily remove them if needed, hoping we won't lost the learnings from this iteration).

**Review by oliviertassinari — APPROVED** [review]

**Comment by oliviertassinari** [comment]
Ohh my bad, I approved too quickly. The visual regression tests report a regression with `docs-components-data-grid-rendering/AutoHeightGrid.png`.

Screenshot 2021-03-25 at 17 32 02

https://www.argos-ci.com/mui-org/material-ui-x/builds/368
